staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation.go

```diff
@@ -0,0 +1,188 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validation
+
+import (
+	"github.com/go-openapi/spec"
+	"github.com/go-openapi/strfmt"
+	"github.com/go-openapi/validate"
+
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+)
+
+func ValidateCustomResource(customresource interface{}, schema *spec.Schema) error {
+	if err := spec.ExpandSchema(schema, nil, nil); err != nil {
+		return err
+	}
+
+	validator := validate.NewSchemaValidator(schema, nil, "spec", strfmt.Default)
+	result := validator.Validate(customresource)
+	if result.AsError() != nil {
+		return result.AsError()
+	}
+	return nil
+}
+
+func ConvertToOpenAPITypes(in *apiextensions.CustomResourceDefinition, out *spec.Schema) error {
+	if err := convertJSONSchemaProps(in.Spec.Validation.JSONSchema, out); err != nil {
+		return err
+	}
+	return nil
+}
+
+func convertJSONSchemaProps(in *apiextensions.JSONSchemaProps, out *spec.Schema) error {
+	if in != nil {
```

Comment:
can we `return nil` early when `in == nil` holds? Less indention is usually better, or at least more idiomatic in golang.